dht.go

```diff
@@ -352,6 +359,44 @@ func makeDHT(ctx context.Context, h host.Host, cfg dhtcfg.Config) (*IpfsDHT, err
 	return dht, nil
 }
 
+// lookupCheck performs a lookup request to a remote peer.ID, verifying that it is able to
+// answer it correctly
+func (dht *IpfsDHT) lookupCheck(ctx context.Context, p peer.ID) error {
+	// lookup request to p requesting for its own peer.ID
+	peerids, err := dht.protoMessenger.GetClosestPeers(ctx, p, p)
+	// p should return at least its own peerid
+	if err == nil && len(peerids) == 0 {
+		return fmt.Errorf("peer %s failed to return its closest peers, got %d", p, len(peerids))
+	}
+	return err
+}
+
+// verifyAndLogPeerQuery returns true if p has been queried less than dht.lookupCheckInterval ago.
+// If the peer has not been queried recently, log that it is being queried.
+func (dht *IpfsDHT) verifyAndLogPeerQuery(p peer.ID) bool {
+	dht.recentlyCheckedPeersLk.Lock()
+	defer dht.recentlyCheckedPeersLk.Unlock()
+
+	now := time.Now()
+
+	// clean recentlyCheckedPeers
+	for peerid, t := range dht.recentlyCheckedPeers {
+		// remove peers that have been queried more than lookupCheckInterval ago
+		if t.Add(dht.lookupCheckInterval).Before(now) {
+			delete(dht.recentlyCheckedPeers, peerid)
+		}
+	}
```

Comment:
Iterating maps is not optimized, this is an expensive operation, you can't iterate the whole map on each query.

I think you can reuse the table refresh, basically instead of scanning the whole routing table you just check if `p` is too old and delete it in that case.